lldb/docs/use/tutorials/writing-custom-commands.md

```diff
@@ -0,0 +1,431 @@
+choices:
+
+- `True`: the completion was handled with no completions.
+
+- `False`: the completion was not handled, forward it to the regular
+completion machinery.
+
+- A dictionary with the key: "completion": there is one candidate,
+whose value is the value of the "completion" key.  Optionally you can pass a
+"mode" key whose value is either "partial" or "complete".  Return partial if
+the "completion" string is a prefix for all the completed value.
+
+For instance, if the string you are completing is "Test" and the available completions are:
+"Test1", "Test11" and "Test111", you should return the dictionary:
+
+```python3
+return {"completion": "Test1", "mode" : "partial"}
+```
+
+and then lldb will add the "1" at the cursor and advance it after the added string,
+waiting for more completions.  But if "Test1" is the only completion, return:
+
+```python3
+{"completion": "Test1", "mode": "complete"}
+```
+
+and lldb will add "1 " at the cursor, indicating the command string is complete.
+
+The default is "complete", you don't need to specify a "mode" in that case.
+
+- A dictionary with the key: "values" whose value is a list of candidate completion
+strings.  The command interpreter will present those strings as the available choices.
+You can optionally include a "descriptions" key, whose value is a parallel array
+of description strings, and the completion will show the description next to
+each completion.
+
+### Loading Commands
+
+One other handy convenience when defining lldb command-line commands is the
+command "command script import" which will import a module specified by file
+path, so you don't have to change your PYTHONPATH for temporary scripts. It
+also has another convenience that if your new script module has a function of
+the form:
+
+```python
+def __lldb_init_module(debugger, internal_dict):
+    # Command Initialization code goes here
+```
+
+where debugger and internal_dict are as above, that function will get run when
+the module is loaded allowing you to add whatever commands you want into the
+current debugger. Note that this function will only be run when using the LLDB
+command `command script import`, it will not get run if anyone imports your
+module from another module.
+
+Another way to load custom commands in lldb is to use the
+`@lldb.command(command_name=None, doc=None)` decorator.
+
+```python3
+@lldb.command()
+def goodstuff(debugger, command, ctx, result, internal_dict):
+    """command help string"""
+    # Command Implementation code goes here
+```
+
+### Examples
+
+Now we can create a module called ls.py in the file ~/ls.py that will implement
+a function that can be used by LLDB's python command code:
+
+```python3
+#!/usr/bin/env python3
+
+import lldb
+import subprocess
+
+def ls(debugger, command, result, internal_dict):
+    proc = subprocess.run(["/bin/ls"] + command.split(), capture_output=True, text=True)
+    result.PutCString(proc.stdout)
```

Comment:
I always print to the `result` as a file, and have assumed that to be the idiomatic way to print to the `result` in python.

```suggestion
    output = subprocess.check_output(["/bin/ls"] + command.split(), text=True)
    print(output, file=result)
```